### PR TITLE
Part 2 RFC comments update

### LINKED ISSUE #530: PublicComment 1.1: Variables in channel?
I'm currently reading through https://docs.ogc.org/DRAFTS/23-057.html as part of the RFC.

In section 8.2.4 I'm wondering how the variables in the channel properties work. There's an example that uses {collectionId}, but I'm unsure where the variables are defined etc. Or are these channels somehow defined externally and are "hardcoded" values that e.g. refer back to specific endpoints? I think this should be explained better. There's not a lot of info on channels in 8.2 (seems to come in 9, but that's after 8.2 so leaves the reader confused).

---

Looking a bit into the future: I think this would also be great extension for STAC API, which could be built on top of EDR PubSub. Does the specification allow extensions, e.g. that a potential channel in STAC could be a specific search query such as `/search?collections=A,B,C&datetime=2020-01-01/..`?

### LINKED ISSUE #531: Public Comment 1.2: UUID/GUID in 10.1.2
I'm currently reading through https://docs.ogc.org/DRAFTS/23-057.html as part of the RFC.

10.1.2 says:
> A universally unique identifier of the message using the **UUID** standard (RFC4122)
> [...]
> Requirement 4 [...] A: A Pub/Sub notification message SHALL provide an id property as a **GUID**.

1. It would be great to link to the RFCs throughout the document.
2. As long as the IDs are unique, does the standard really need to define how the IDs look like?
3. GUID is not mentioned before. What is it? Is it the same as a UUID? There are also more mentions of GUID in the annex.

### LINKED ISSUE #532: Public Comment 1.3: Allowed values for operation
I'm currently reading through https://docs.ogc.org/DRAFTS/23-057.html as part of the RFC.

Chapter 10.1.2.2 says:

> The operation property indicates the stage of the lifecycle for the resource described in the notification, and can be used to notify users that a resource has been updated or deleted. If not specified, the default value is create.

1. The first sentence seems to imply update and delete, but then in the next sentence the default is create. Shouldn't this be "that a resource has been added, updated or deleted"? The same applies to Req. 6 A
2. You should provide a list of allowed values in the description. I see "delete" (example) and "create" (default value). I assume for updated the value is "update" but that's not specified.

Permission 3 says:

> A Pub/Sub Notification Message MAY provide the properties.operation property to indicate if a resource has been inserted.

What does this mean? How is insertion different from created?

### LINKED ISSUE #533: Public Comment 1.4: Collections, (Geo)JSON and Properties
I'm currently reading through https://docs.ogc.org/DRAFTS/23-057.html as part of the RFC.

This is more a set of high-level thoughts and questions.

- It looks like channel / is "all notifications". Is this only items or e.g. also new collections? Is there a way to only be notified about new collections?
- It sounds like GeoJSON is one of the payloads, but not necessarily required. So could this also be a Collection JSON?
- If yes, it is a bit weird that the properties (which originate from GeoJSON Features) is explicitly referenced in 10.1.2.x for pubtime and operation. The Collections don't have properties. That seems to bei GeoJSON specific.

### LINKED ISSUE #534: Advertising HTML-encoded AsyncAPI definition documents in PubSub enabled OGC API landing pages
This issue is related to https://github.com/opengeospatial/ogcapi-common/issues/329 , however is more severe for HTML-encoded definition documents.

Consider the following fragment from the links in a landing page.

```json
        {
            "rel": "service-doc",
            "type": "text/html",
            "title": "The OpenAPI definition as HTML",
            "href": "http://localhost:5000/openapi?f=html",
            "hreflang": "en-US"
        }
```

If a server wishes to advertise links to HTML-encoded AsyncAPI definition documents and HTML-encoded OpenAPI definition documents, there is a need to either provide a `rel` or `type` that allows such distinction to be made. Any thoughts on how this could be achieved?

Note that the reason why the issue is more severe for HTML-encoded definition documents, is because at least for JSON-encoded definition documents there is a specific media type for OpenAPI definition documents.

### LINKED ISSUE #535: Public Comment 1.5: x-ogc-api-link object in /req/pubsub-channels/channels is unclear
EDR, Part 2, RFC version:

`/req/pubsub-channels/channels` mentions "an `x-ogc-api-link` object", but this object is undefined and only used in examples. Even from the examples, I do not understand the use of the `rel` and `type` members.

The requirement should either be deleted or it needs much more detail to be meaningful - and testable.

### LINKED ISSUE #537: Public Comment 1.7: Message payload requirements
EDR, Part 2, RFC version:

`/req/pubsub-message-payload/id` requires that a different id is used for the message than the identifier of the resource. Since `/req/pubsub-channels/channels` relates the channels to resources of the API, which all have IDs, another option would be to use the resource id also for the payload id. This approach seems more meaningful, for example, for features. The use in EDR Part 2 is fine, if it is not understood to be a general rule for Pub/Sub payloads of other OGC API resources.

Similar concerns will apply to the other requirements in the requirements class. For example, a feature may already have `pubtime` property.

### PR DESCRIPTION
**Pull request by tomkralidis**
WIP PLEASE DO NOT MERGE.

- fixes #530
- fixes #531
- fixes #532
- fixes #533
- fixes #534 
- fixes #535
- fixes #537
- makes requirements class `pubsub` a dependency of requirements class `pubsub-channels`
- makes conformance class `pubsub` a dependency of conformance class `pubsub-channels`

### CODE REVIEW

File: extensions/pubsub/standard/sections/clause_9_pubsub-message-payload.adoc [hunk 1]
```diff
@@ -29,7 +29,8 @@ include::../recommendations/pubsub-message-payload/REC_crs.adoc[]
 ==== Identifier
 
 A universally unique identifier of the message using the UUID standard (RFC4122). The identifier is generated by the originator of the message.
-It remains the same throughout the lifetime of the message.
+
+It remains the same throughout the lifetime of the message.  Note that multiple messages may communicate updates on the same resource.  The identifier is also valuable for "replay" feed services, where a Pub/Sub client may have gone offline for a period of time.
```
Comment (m-mohr) [inline extensions/pubsub/standard/sections/clause_9_pubsub-message-payload.adoc:33]: > Note that multiple messages may communicate updates on the same resource. 

What's the consequence of this? Is there any? Any recommended behaviour?

Comment (tomkralidis) [reply extensions/pubsub/standard/sections/clause_9_pubsub-message-payload.adoc]: > > Note that multiple messages may communicate updates on the same resource.
> 
> What's the consequence of this? Is there any? Any recommended behaviour?

I've added a recommendation for `properties.resource_id` in this context.

File: extensions/pubsub/standard/requirements/pubsub-geojson-fg-json-message-payload/REQ_rc-id.adoc
Comment (m-burgoyne) [inline extensions/pubsub/standard/requirements/pubsub-geojson-fg-json-message-payload/REQ_rc-id.adoc:1]: Should this be  A Pub/Sub notification message SHALL provide an `+id+` property as a globally unique identifier for the message ?

**Review by m-burgoyne — APPROVED** [review]

File: extensions/pubsub/standard/requirements/pubsub-geojson-fg-json-message-payload/REQ_rc-id.adoc
Comment (tomkralidis) [reply extensions/pubsub/standard/requirements/pubsub-geojson-fg-json-message-payload/REQ_rc-id.adoc:1]: Fixed

**Review by m-burgoyne — APPROVED** [review]

File: extensions/pubsub/standard/sections/annex-pubsub.adoc [hunk 2]
```diff
@@ -142,6 +142,8 @@ Different Pub/Sub protocols are supported as additional server objects, and can
 ** `collections/wthr_stn/items`
 ** `collections/stream_gage/items`
 
+The use of `x-ogc-api-link` demonstrates the capability to cross reference an AsyncAPI to an OGC API resource.
```
Comment (cportele) [inline extensions/pubsub/standard/sections/annex-pubsub.adoc:145]: The same concerns as in #451 apply here as well.

`rel` is used incorrectly and are not consistent with the definitions of the link relation type. The link relations `collection` or `items` are not applicable to the context of the link (a channel).

The use of `type` is also unclear to me.

My guess is that this construct is intended to identify a resource where a change in the resource results in a message and that also the type of the resource should be declared. The current approach as I understand it from the examples and that still is not specified, does not provide a solution for this.

File: extensions/pubsub/standard/sections/clause_7_pubsub.adoc [hunk 3]
```diff
@@ -39,18 +39,22 @@ Based on research and testing, the Pub/Sub White Paper recommended the use of As
...
+NOTE: the media type for an AsyncAPI document may change in the future as decided by the AsyncAPI community.
+
 include::../requirements/pubsub/REQ_rc-api.adoc[]
 
 ==== Providing notification metadata as an OGC API endpoint
 
-For Brokers providing notification metadata (as opposed to actual data payloads), an implementation of OGC API Building Blocks can, in parallel, readily provide GeoJSON-based notification messages via an OGC API - Features endpoint. Providing message payloads via an implementation of OGC API Standard(s) provides the additional benefit of querying for past messages over time in case of a lost connection.
+For Brokers providing notification metadata (as opposed to actual data payloads), an implementation of OGC API Building Blocks can, in parallel, readily provide GeoJSON-based notification messages via an OGC API - Features endpoint. Providing message payloads via an implementation of OGC API Standard(s) provides the additional benefit of querying for past messages over time in case of a lost connection.  See <<pubsub-geojson-fg-json-message-payload-section>> for more information.
 
 ==== Providing Pub/Sub links to collection updates
```
Comment (cportele) [inline extensions/pubsub/standard/sections/clause_7_pubsub.adoc:50]: My comment in #451 still applies. These links are not web links. The link will appear broken and useless to a client that understands web links (resolving the `href` will not return a resource in the stated media type or that meets the semantics of the link relation type). I think EDR Part 2 should not represent this information in that way and a different approach should be used to represent this information.

If the EDR SWG decides to follow the current approach, at least remove the Features examples. Examples and requirements for Features should be in the scope of the Features API SWG (and I do not see that such an approach would be used for those resources).

File: extensions/pubsub/standard/sections/clause_8_pubsub-channels.adoc [hunk 4]
```diff
@@ -16,5 +16,6 @@ For example, users could use a subscription to metadata records, which are usual
 
 Using the OGC API service endpoints of the URL path of resource and sub-resources provides the key benefit that developers implementing OGC API Standards do not need to learn a different, additional approach or resource path for Pub/Sub (same content, additional interface).
 
+While AsyncAPI does not have native OpenAPI integration, in order to cross-reference a given AsyncAPI channel against its associated OGC API resource, the `+x-ogc-api-link+` object can be used.  The structure of this object is defined in OGC API - Common clause 8.1.
```
Comment (cportele) [inline extensions/pubsub/standard/sections/clause_8_pubsub-channels.adoc:19]: OGC API - Common clause 8.1 is "HTTP 1.1". I assume this should be clause 6.3 ("Links")?

**Review by cportele — CHANGES_REQUESTED** [review]
See the specific comments.

In addition, the PR states that it closes #537. It only does that, if it is clarified either in the issue or in a note in the document that the payload related requirements classes are applicable in the context of EDR, but likely not applicable in other contexts (e.g., when communicating changed representations of a Feature or Collection resource).

File: extensions/pubsub/standard/sections/clause_9_pubsub-geojson-fg-json-message-payload.adoc [hunk 5]
```diff
@@ -53,14 +56,34 @@ The ``pubtime`` property identifies the date/time of when the message was posted
...
 ----
 
-include::../requirements/pubsub-message-payload/REQ_rc-pubtime.adoc[]
+include::../requirements/pubsub-geojson-fg-json-message-payload/REQ_rc-pubtime.adoc[]
 
 ===== operation
 
 The ``operation`` property indicates the stage of the lifecycle for the resource described in the notification,
-and can be used to notify users that a resource has been updated or deleted. If not specified, the default value is ``create``.
+and can be used to notify users that a resource has been created, updated or deleted. If not specified, the default value is ``create``.
```
Comment (m-mohr) [inline extensions/pubsub/standard/sections/clause_9_pubsub-geojson-fg-json-message-payload.adoc:64]:
```suggestion
and can be used to notify users that a resource has been created, updated or deleted. If not specified, the default value is ``create``. Other allowed values are ``update`` and ``delete``.
```

Comment (tomkralidis) [reply extensions/pubsub/standard/sections/clause_9_pubsub-geojson-fg-json-message-payload.adoc]: Fixed.

**Review by chris-little — APPROVED** [review]
This addresses all the comments (we think!) but we need to see a final document to understand in context.

**Comment by chris-little** [comment]
As agreed at EDR API SWG 116, we would merge so final document can be reviewed.
@tomkralidis Can you make sure that a PDF is created. My workflow always crashes for unknown reason (probably wrong permissions).

**Comment by tomkralidis** [comment]
FYI a rendered PDF now exists in [document.pdf](https://github.com/opengeospatial/ogcapi-environmental-data-retrieval/files/15263543/document.pdf).

**Pull request merged**